internal/controller/controllers/agent_controller.go

```diff
@@ -204,6 +210,132 @@ func (r *AgentReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (
 	return r.updateStatus(ctx, log, agent, &h.Host, h.ClusterID, nil, false)
 }
 
+func findMatchingNode(agent *aiv1beta1.Agent, nodes []corev1.Node) *corev1.Node {
+	addresses := make([]string, 0)
+	for _, intf := range agent.Status.Inventory.Interfaces {
+		for _, ip := range append(intf.IPV4Addresses, intf.IPV6Addresses...) {
+			splitIp := strings.Split(ip, "/")
+			if len(splitIp) == 2 {
+				addresses = append(addresses, splitIp[0])
+			}
+		}
+	}
+	for i := range nodes {
+		n := &nodes[i]
+		if strings.EqualFold(n.Name, agent.Spec.Hostname) {
+			return n
+		}
+		for _, addr := range n.Status.Addresses {
+			if addr.Type == corev1.NodeInternalIP && funk.ContainsString(addresses, addr.Address) {
+				return n
+			}
+		}
+	}
+	return nil
+}
+
+func isNodeReady(node *corev1.Node) bool {
+	for _, c := range node.Status.Conditions {
+		if c.Status == corev1.ConditionTrue && c.Type == corev1.NodeReady {
+			return true
+		}
+	}
+	return false
+}
+
+func isNodeApproved(agent *aiv1beta1.Agent, nodes []corev1.Node) bool {
+	if agent.Status.DebugInfo.State == models.HostStatusAddedToExistingCluster &&
+		agent.Status.Progress.CurrentStage == models.HostStageRebooting {
+		node := findMatchingNode(agent, nodes)
+		if node != nil {
+			return isNodeReady(node)
+		}
+	}
+	return false
+}
+
+func (r *AgentReconciler) shouldApproveCSRs(agent *aiv1beta1.Agent, clients ClusterKubeClient) (bool, error) {
+
+	nodes, err := clients.ListNodes()
+	if err != nil {
+		return false, err
+	}
+	return !isNodeApproved(agent, nodes.Items), nil
+}
+func isCsrApproved(csr *certificatesv1.CertificateSigningRequest) bool {
+	for _, c := range csr.Status.Conditions {
+		if c.Type == certificatesv1.CertificateApproved {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *AgentReconciler) isAssociatedWithAgent(csr *certificatesv1.CertificateSigningRequest, agent *aiv1beta1.Agent) (bool, error) {
+	decodedCSR, _ := pem.Decode(csr.Spec.Request)
+	if decodedCSR == nil {
+		return false, errors.Errorf("Failed to decode request of CSR %s", csr.Name)
+	}
+	cr, err := x509.ParseCertificateRequest(decodedCSR.Bytes)
+	if err != nil {
+		return false, errors.Wrapf(err, "Failed to parse CSR %s", csr.Name)
+	}
+	return strings.EqualFold("system:node:"+agent.Spec.Hostname, cr.Subject.CommonName), nil
+}
+
+func (r *AgentReconciler) approveAIHostsCSRs(clients ClusterKubeClient, agent *aiv1beta1.Agent) error {
+	csrList, err := clients.ListCsrs()
+	if err != nil {
+		r.Log.WithError(err).Error("Failed to get CSRs")
+		return err
+	}
+	for i := range csrList.Items {
+		csr := &csrList.Items[i]
+		if !isCsrApproved(csr) {
+			associated, err := r.isAssociatedWithAgent(csr, agent)
+			if err != nil || !associated {
+				if err != nil {
+					r.Log.WithError(err).Error("Find agent <-> CSR association")
+				}
+				continue
+			}
+			if err = clients.ApproveCsr(csr); err != nil {
+				r.Log.WithError(err).Error("Failed to approve CSR")
```

Comment:
can you add more info to this error? What agent? what node? what cluster?

Reply:
OK